### PR TITLE
Upload command

### PR DESCRIPTION
**Pull request by saeta**
The upload commands lets instructional teams upload their graders directly into Coursera's production environment. The uploaded graders become a draft, that the instructional team can then publish in their draft through the authoring interface on Coursera.org

In order to integrate with Coursera, the `courseraprogramming` tool uses Coursera's published OAuth2 implementation at: https://tech.coursera.org/app-platform/oauth2/

This pull request also adds a semi-reusable OAuth2 client library.

### CODE REVIEW

File: courseraprogramming/commands/config.py [hunk 1]
```diff
@@ -20,15 +20,77 @@
 You may install it from source, or via pip.
 """
 
+from courseraprogramming.commands import oauth2
```
Comment (bryankane) [inline courseraprogramming/commands/config.py:23]: pep8 says that local library imports should be last: http://legacy.python.org/dev/peps/pep-0008/#imports

File: courseraprogramming/commands/config.py [hunk 2]
```diff
@@ -20,15 +20,77 @@
...
+    if not args.quiet > 0:
+        print 'Name: %s' % name
+        print 'External ID: %s' % external_id
+
+    if name is None or external_id is None:
+        sys.exit(1)
+
+
+def display_auth_cache(args):
+    '''
```
Comment (bryankane) [inline courseraprogramming/commands/config.py:71]: you're inconsistent in your docstring format. pick either single or double quotes. (pep8 says double quotes : http://legacy.python.org/dev/peps/pep-0008/#documentation-strings)

File: courseraprogramming/commands/oauth2.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+        return 'Attempt to use expired token. Expired at: %s.' % \
+            self.oauth2Auth.expires
+
+
+class OAuth2Exception(Exception):
+    def __init__(self, msg):
+        self.msg = msg
+
+    def __str__(self):
+        return 'OAuth2 Protocol Exception: %(msg)s' % {
```
Comment (bryankane) [inline courseraprogramming/commands/oauth2.py:50]: you're inconsistent in your use of named parameters / parameters by order in string interpolation.

File: courseraprogramming/commands/oauth2.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+            logging.debug('Local cache is good.')
+        return CourseraOAuth2Auth(self.token_cache['token'],
+                                  self.token_cache['expires'])
+
+
+def build_oauth2(args, cfg=None):
+    if cfg is None:
+        cfg = configuration()
+
+    try:
```
Comment (bryankane) [inline courseraprogramming/commands/oauth2.py:440]: this feels very weird to be using try / excepts here. can't you do

```
if client_id in args:
    client_id = args.client_id
else:
    client_id = cfg.get('oauth2', 'client_id')
```

File: courseraprogramming/commands/upload.py [hunk 5]
```diff
@@ -21,15 +21,273 @@
...
+    Saves the container image to the file system in tar form. (similar to the
+    `docker save` command.)
+
+    Returns the name of the file containing the export.
+    '''
+    # TODO: get information on the image, and run a few basic sanity checks.
+    # (e.g. check for ENTRYPOINT, etc.)
+
+    image = d.get_image(args.containerId)
+    image_file_name = \
```
Comment (bryankane) [inline courseraprogramming/commands/upload.py:54]: `image_file_name = args.file_name or args.containerId`?

Comment (bryankane) [inline courseraprogramming/commands/upload.py:54]: (no idea if that works. but something like that?)

File: courseraprogramming/commands/oauth2.py (hunk 4 above)
Comment (saeta) [inline courseraprogramming/commands/oauth2.py:440]: I am using the [EAFP](https://docs.python.org/2/glossary.html#term-eafp) pattern, which is apparently the preferred approach.

**Pull request merged**